### ISSUE TITLE
[Regression]: Setting code-style options severity via `Tools` → `Options` has no effect

### ISSUE THREAD

**Issue by Youssef1313** [dotnet/roslyn#61844]
Visual Studio 2019:

![image](https://user-images.githubusercontent.com/31348972/173340099-22742251-e42a-4452-95de-5c475f47102c.png)

---

Visual Studio 2022 17.2.0 Preview 1:

![image](https://user-images.githubusercontent.com/31348972/173340388-3e4df9e1-df4d-4119-81a1-0aa12f1be778.png)

@tmat I'm not sure if this is a result of the recent refactoring.

**Comment by Youssef1313** [comment]
Taking a random example code-style options:

https://github.com/dotnet/roslyn/blob/a528fe2d1721466c4ada3e97bd99a70521560f58/src/Analyzers/CSharp/Analyzers/CodeStyle/CSharpAnalyzerOptionsProvider.cs#L98-L101

Are we always reading from editorconfig and ignoring roaming location?

**Comment by tmat** [comment]
Possibly. I'll take a look.

**Comment by tmat** [comment]
I see issues with diagnostics being not reported or reported multiple times:

https://user-images.githubusercontent.com/41759/176246717-9aa4e250-8f78-44d1-921b-bc92e6fc9c80.mp4

The option value seems to be correctly passed to the analyzer, but the analyzer is not always invoked as I would expect.

**Comment by tmat** [comment]
@dibarbet

**Comment by tmat** [comment]
@dibarbet 
![image](https://user-images.githubusercontent.com/41759/176253318-a7eec4c0-876c-4deb-b977-96d839244c7b.png)

**Comment by tmat** [comment]
One thing that does not work with LSP - changing the severity from error to warning does not refresh the diagnostics right away, only if I change the file. Maybe we need to trigger pull on global option change?

**Comment by tmat** [comment]
Interesting, even closing and reopening the document after the severity is changed does not update the LSP diag.

**Comment by tmat** [comment]
> Sorry, what do you mean by that?

The code that reads the option is not executed.

**Comment by dibarbet** [comment]
> @dibarbet It works fine if I turn on LSP.

Good to know!

> One thing that does not work with LSP - changing the severity from error to warning does not refresh the diagnostics right away, only if I change the file. Maybe we need to trigger pull on global option change?

Ah this is interesting.  The client polls us, but we're looking at the solution snapshot to see if anything has changed before attempting to compute diagnostics.   This doesn't contain global options so even when they change we don't recompute diagnostics.  Is there a global option version / checksum we can cache / look at to see if they've changed since the last request?  I could also listen to the option changed event on it but that is a bit messier.

**Comment by tmat** [comment]
Also seeing contract failure when displaying Tools Options under debugger.
```
 	Microsoft.CodeAnalysis.Workspaces.dll!Roslyn.Utilities.Contract.ThrowIfNull<System.Threading.Tasks.Task>(System.Threading.Tasks.Task value, int lineNumber) Line 29	C#
>	Microsoft.CodeAnalysis.EditorFeatures.dll!Microsoft.CodeAnalysis.Editor.Shared.Preview.PreviewSolutionCrawlerRegistrationServiceFactory.Service.UnregisterAsync(Microsoft.CodeAnalysis.Workspace workspace) Line 119	C#
 	Microsoft.CodeAnalysis.EditorFeatures.dll!Microsoft.CodeAnalysis.Editor.Shared.Preview.PreviewSolutionCrawlerRegistrationServiceFactory.Service.Unregister(Microsoft.CodeAnalysis.Workspace workspace, bool blockingShutdown) Line 114	C#
 	Microsoft.CodeAnalysis.EditorFeatures.dll!Microsoft.CodeAnalysis.Editor.Shared.Preview.PreviewWorkspace.Dispose(bool finalize) Line 111	C#
 	Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.Workspace.Dispose() Line 353	C#
 	Microsoft.VisualStudio.LanguageServices.Implementation.dll!Microsoft.VisualStudio.LanguageServices.Implementation.Options.AbstractOptionPreviewViewModel.UpdatePreview.AnonymousMethod__2(object s, System.EventArgs a) Line 168	C#
 	Microsoft.VisualStudio.Platform.VSEditor.dll!Microsoft.VisualStudio.Text.Utilities.GuardedOperations.RaiseEvent(object sender, System.EventHandler eventHandlers)	Unknown
 	Microsoft.VisualStudio.Platform.VSEditor.dll!Microsoft.VisualStudio.Text.Editor.Implementation.WpfTextViewHost.Close()	Unknown
 	Microsoft.VisualStudio.LanguageServices.Implementation.dll!Microsoft.VisualStudio.LanguageServices.Implementation.Options.AbstractOptionPreviewViewModel.Dispose() Line 207	C#
 	Microsoft.VisualStudio.LanguageServices.Implementation.dll!Microsoft.VisualStudio.LanguageServices.Implementation.Options.GridOptionPreviewControl.Close() Line 112	C#
 	Microsoft.VisualStudio.LanguageServices.Implementation.dll!Microsoft.VisualStudio.LanguageServices.Implementation.Options.AbstractOptionPage.OnClosed(System.EventArgs e) Line 137	C#

```

**Comment by CyrusNajmabadi** [comment]
@tmat is there anything remaining to be done here?  if not, can you close this out?
